### PR TITLE
Fixes various get_bgp_neighbors

### PR DESCRIPTION
**Pull request by tehhobbit**
<!-- Make sure you have read http://napalm.readthedocs.io/en/latest/contributing/index.html --!>

- The use of a for loop instead of while caused random data loss
  ( the actual cause of issue #58)<br>

- Peers that where idle would always have enable=False<br>

- Uptime was actually the timestamp of when the session started<br>

 - Use precompiled regex and more generic approach instead of a load of private methods<br>

<br><br>
This PR have 2 failing tests that might take some discussion to solve
<br><br>

1) Issue #58 is actually caused by the use of a for loop and the test data is invalid<br>
2) Since uptime is now calculated and not static, making the moc data wrong. The one way I came up with was mock time.time() in napalm-base, but this might have effects on other drivers so think it warrants some more thinking. <br>

### CODE REVIEW

File: napalm_eos/eos.py
Comment (tehhobbit) [inline napalm_eos/eos.py:369]: This is the actual reason for issue #58

**Comment by tehhobbit** [comment]
I'm available on slack since this is a bit of a weird one

**Comment by dbarrosop** [comment]
1. Looks like tests are not passing.
2. Could you add a test case covering the bug you are fixing?

**Comment by tehhobbit** [comment]
The issue 58 test case isnt passing because the mock data is simply wrong, there is no case when the data would be missing from 'show ip bgp summary|json' just because the session is  down ( checked with several releases) That one I can easily fix by fixing the mock data.

The second test is a bit more complicated. The bug is that this function have always returned start time of bgp session rather than the uptime. Fixing this leads to a problem with the static mock data since the value is calculated based on system clock when the tests are run. One way is to patch time.time() to return a static value, but due to how napalms tests work I would have to patch it in napalm_base. So wanted input from maintainers before doing that.

**Comment by mirceaulinic** [comment]
> The issue 58 test case isnt passing because the mock data is simply wrong, there is no case when the data would be missing from 'show ip bgp summary|json' just because the session is down

Want a screenshot?

**Comment by tehhobbit** [comment]
Not sure I follow, and I could not find a case where the sessions was missing from 'show ip bgp summary' but would show up in 'show ip bgp neighbors'. The bug that was there was the combination of a for loop and popping elements https://asmeurersympy.wordpress.com/2009/07/20/modifying-a-list-while-looping-through-it-in-python/. But if I am missing something here I'd put that old code there as a precaution

**Comment by mirceaulinic** [comment]
> Not sure I follow, and I could not find a case where the sessions was missing from 'show ip bgp summary' but would show up in 'show ip bgp neighbors'.

It does not make sense to be there indeed. But believe it or not, it does exist.

> The bug that was there was the combination of a for loop and popping elements https://asmeurersympy.wordpress.com/2009/07/20/modifying-a-list-while-looping-through-it-in-python/. But if I am missing something here I'd put that old code there as a precaution

Thanks for your fix!

https://github.com/napalm-automation/napalm-eos/pull/69 come as a fix for that, but I did not dig into it very much. I will analyse and test it more carefully tomorrow!

Meanwhile:

> I'm available on slack since this is a bit of a weird one.

Can you be more specific? As you've expressed your desire to refactor it, feel free to come with a valid proposal, even if very different, that passes the tests.

**Comment by mirceaulinic** [comment]
For the record:

```
#sh ipv6 bgp neighbors <hidden_address>
BGP neighbor is <hidden_address>, remote AS <hidden_as>, external link
  BGP version 4, remote router ID 0.0.0.0, VRF default
  Negotiated BGP version 0
  Last read never, last write never
  Hold time is 180, keepalive interval is 60 seconds
  Configured hold time is 180, keepalive interval is 60 seconds
  Connect timer is inactive
  Idle-restart timer is inactive
  BGP state is Idle, Peer is not activated in any address-family mode
  Peering failure hint: Peer is not activated in any address-family mode
  Number of transitions to established: 0
  Last state was NoState
  Last event was NoEvent
  Neighbor Capabilities:
    Four Octet ASN: advertised
    Route Refresh: advertised
    Send End-of-RIB messages: advertised
```

Retrieving the summary:

```
#sh ipv6 bgp summary vrf  all | i <hidden_address>
```

Is not there.

Is this clear now @tehhobbit?

**Review by mirceaulinic — CHANGES_REQUESTED** [review]
Please make it pass the tests, also for situations such #58.

**Comment by dbarrosop** [comment]
@tehhobbit 

>  I could not find a case where the sessions was missing from 'show ip bgp summary' but would show up in 'show ip bgp neighbors'

Does @mirceaulinic's example make sense for you? Just wanted to make sure we are all talking about the same

> The second test is a bit more complicated. The bug is that this function have always returned start time of bgp session rather than the uptime

Is that true? It's interesting we didn't catch it earlier. The testing framework has a way of dealing with that, the value in the expected result can be `"..."` and the testing framework will interpret that as a value that might change due to reasons and just check the type. For example: https://github.com/napalm-automation/napalm-eos/blob/develop/test/unit/mocked_data/test_get_facts/normal/expected_result.json#L14

**Comment by tehhobbit** [comment]
I have sorted the uptime. But looking at the mockdata for the case of missing output (Davids example) it makes little to sense to me to have it return the address_family key given that the reason for the output missing in 'show ip bgp summary' is that address family is actually not properly configured.

**Comment by dbarrosop** [comment]
> I have sorted the uptime

Great

> (Davids example) it makes little to sense to me to have it return the address_family

Which example? I haven't provided any example : ) Are you referring to the issue58?

**Comment by tehhobbit** [comment]
Yeah I was referring to issue58

**Comment by mirceaulinic** [comment]
Erm... Yes, would be an empty dictionary, as there won't be any counters available. So your logic says that if the leaves are missing to cut the tree, right? Anyway, if you do that, the tests will fail as the ```address_family``` key must have a dict value.

**Comment by dbarrosop** [comment]
Ok, I am not sure I understand the problem. @mirceaulinic how can I reproduce? I have been testing with 4.15.1M and a BGP session down seems to return data:

```
localhost#show ipv6 bgp summary vrf all | json
{
    "vrfs": {
        "default": {
            "routerId": "10.0.2.15",
            "peers": {
                "2001:db8::2": {
                    "msgSent": 0,
                    "inMsgQueue": 0,
                    "prefixReceived": 0,
                    "upDownTime": 1480325830.250048,
                    "version": 4,
                    "msgReceived": 0,
                    "prefixAccepted": 0,
                    "peerState": "Active",
                    "outMsgQueue": 0,
                    "underMaintenance": false,
                    "asn": 65002
                }
            },
            "vrf": "default",
            "asn": 65001
        }
    }
}
```

Or am I misunderstanding something?

**Comment by mirceaulinic** [comment]
Does the output of ```sh ip bgp neighbors vrf all``` look the same too? It must be a  session down, that never started.

**Comment by dbarrosop** [comment]
The BGP session was never up, it's the only VM I had on my test :P

I didn't enable the ipv4 address family so that one was empty.

What I noticed is the following though:

```
localhost#sh run | s router bgp
router bgp 65001
   neighbor 2001:db8::2 remote-as 65002
   neighbor 2001:db8::2 maximum-routes 12000
   address-family ipv4
      neighbor 2001:db8::2 activate
   !
   address-family ipv6
      no neighbor 2001:db8::2 activate
localhost#show ipv6 bgp neighbors
BGP neighbor is 2001:db8::2, remote AS 65002, external link
  BGP version 4, remote router ID 0.0.0.0, VRF default
  Negotiated BGP version 0
  Last read never, last write never
  Hold time is 180, keepalive interval is 60 seconds
  Configured hold time is 180, keepalive interval is 60 seconds
  Connect timer is inactive
  Idle-restart timer is inactive
  BGP state is Idle, Peer is not activated in any address-family mode
  Number of transitions to established: 0
  Last state was NoState
  Last event was NoEvent
  Neighbor Capabilities:
    Four Octet ASN: advertised
    Route Refresh: advertised
    Send End-of-RIB messages: advertised
  Restart timer is inactive
  End of rib timer is inactive
  Message statistics:
    InQ depth is 0
    OutQ depth is 0
                         Sent      Rcvd
    Opens:                  0         0
    Notifications:          0         0
    Updates:                0         0
    Keepalives:             0         0
    Route-Refresh:          0         0
    Total messages:         0         0
  Prefix statistics:
                         Sent      Rcvd
    IPv4 Unicast:           0         0
    IPv6 Unicast:           0         0
  Inbound updates dropped by reason:
    AS path loop detection: 0
    Enforced First AS: 0
    Malformed MPBGP routes: 0
    Originator ID matches local router ID: 0
    Nexthop matches local IP address: 0
    Nexthop invalid for single hop eBGP: 0
Local AS is 65001, local router ID 10.0.2.15
TTL is 1
Remote TCP address is 2001:db8::2, remote port is 179
Auto-Local-Addr is disabled

localhost#show ipv6 bgp summary vrf all | json
{
    "vrfs": {
        "default": {
            "routerId": "10.0.2.15",
            "peers": {},
            "vrf": "default",
            "asn": 65001
        }
    }
}
```

Which I am not sure it was your use case but yields the same result so we should accommodate it.

**Comment by mirceaulinic** [comment]
Anyway, @tehhobbit's question is, if I understand correctly: shall we return the key ```address_family``` in a case such #58, or simply strip that key from the output dict?

**Comment by dbarrosop** [comment]
He can clarify but I don't think that's the question. In any case, just to be clear, data has to be always consistent so it has to be returned.

**Comment by dbarrosop** [comment]
@tehhobbit as there are circumstances where a BGP peer might be configured and not show on `show ipv6 bgp summary vrf all` as proved on the example above (which yields the same result as issue58), could you accommodate it, please? I agree it's not a usual case but it's known so it would be nice to handle it rather than breaking the code/test.

Thanks!

File: test/unit/mocked_data/test_get_bgp_neighbors/normal/expected_result.json
Comment (mirceaulinic) [inline test/unit/mocked_data/test_get_bgp_neighbors/normal/expected_result.json:7]: No, please do not change the tests under the ```normal``` test case.
As this is a corner case, not the default behaviour, please create a new test case for the output you see.

**Review by mirceaulinic — CHANGES_REQUESTED** [review]
Do not change the content of the ```expected_result```

File: test/unit/mocked_data/test_get_bgp_neighbors/normal/expected_result.json
Comment (tehhobbit) [reply test/unit/mocked_data/test_get_bgp_neighbors/normal/expected_result.json:7]: This is not a corner case but actually another bug. What arista returns is the timestamp when a session is started, not the uptime of the session, so with the change I now calculate the uptime using time.time() this leads to the uptime being variable

Comment (mirceaulinic) [reply test/unit/mocked_data/test_get_bgp_neighbors/normal/expected_result.json:7]: So... you'll return ```...``` instead of a valid int for the timestamp?

Comment (dbarrosop) [reply test/unit/mocked_data/test_get_bgp_neighbors/normal/expected_result.json:7]: Yes, the "..." is just a workaround for results that may vary depending on when/where the test was run. This is a perfect example. @mirceaulinic check "get_facts" test case for another one.

Comment (mirceaulinic) [reply test/unit/mocked_data/test_get_bgp_neighbors/normal/expected_result.json:7]: Alright then, thanks for info @dbarrosop

**Comment by dbarrosop** [comment]
Thx @tehhobbit for the fixes. I think we might have found a minor issue or two but conceptually the PR looks great now so once we address them we will merge this. @mirceaulinic will get back to you soon-ish : )

**Comment by mirceaulinic** [comment]
@tehhobbit I was testing your branch and hit the error:

```python
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "napalm_eos/eos.py", line 389, in get_bgp_neighbors
    'remote_id': napalm_base.helpers.ip(rid_info.group('rid')),
AttributeError: 'NoneType' object has no attribute 'group'
```

Debugging it, reveals that ```rid_info``` was None for the following case.

```
#sh ip bgp neighbors <hidden> | include remote AS | remote router ID |IPv[46] Unicast:.*[0-9]+|^Local AS|Desc|BGP state
BGP neighbor is <hidden>, remote AS 13335, internal link
  Description: "SIR/pmacct"
  BGP version 4, remote router ID 1.2.3.4, VRF default
  BGP state is Connect
    IPv4 Unicast:           0         0
    IPv6 Unicast:           0         0
Local AS is 13335, local router ID <hidden>
```

So is the case when the neighbor has description:

```python
rid_info = re.match(self._RE_BGP_RID_INFO, lines.pop(9))
```

But ```lines.pop(9)``` is not correct:

```
    IPv6 Unicast:           0         0
```

Please revisit the case when neighbors have descriptions.

I suppose it should be ```lines.pop(0)``` instead.

**Comment by tehhobbit** [comment]
Give me a few minutes and I will sort that one out

**Comment by mirceaulinic** [comment]
I think you only need to change SLOC #381 to ```rid_info = re.match(self._RE_BGP_RID_INFO, lines.pop(0))```

**Comment by mirceaulinic** [comment]
I confirm it works now @tehhobbit!

I would have one more minor request: please do not use defaultdict, as the return would be:

```python
>>> e.get_bgp_neighbors()
defaultdict(<function <lambda> at 0x7fd477ff9c80>, {u'global': { ... }})
>>> type(e.get_bgp_neighbors())
<type 'collections.defaultdict'>
```

The returned datatype should be a dict instead a similar iterable object.

**Comment by tehhobbit** [comment]
I introduced a small helper to provide defaults where things might be missing. But looking at the case with missing address_family this code is still very brittle. If the lines with  

   IPv4 Unicast:           0         0
    IPv6 Unicast:           0         0

are missing, the technique of popping lines will actually cause everything to be offset and therefor yield data that is completely off and very hard to debug.

**Comment by mirceaulinic** [comment]
> If the lines with
IPv4 Unicast: 0 0
IPv6 Unicast: 0 0
are missing, the technique of popping lines will actually cause everything to be offset and therefor yield data that is completely off and very hard to debug.

You don't have to follow the same pattern @tehhobbit - you can, for example use textFSM templates. Or whatever you feel comfortable with & is a smarter way of dealing with that raw text output.

**Comment by tehhobbit** [comment]
Yeah I will not do this in this PR, it was mostly a note that the code could do with a revamp due to these kind of weaknesses

**Comment by mirceaulinic** [comment]
Okay, as you wish :)

If you have the time and happy to contribute again, we'll be waiting for a PR with a better approach.

For the moment, please avoid the usage of ```defaultdict``` (see above).

**Comment by tehhobbit** [comment]
Defaultdict sorted

**Review by mirceaulinic — APPROVED** [review]

**Pull request merged**
